docs: scaling Coder (#5550)
coder/coder: docs/admin/scale/index.md

- **concurrent workspace builds**: Workspace operations (e.g. create/stop/delete/apply) across all users
- **concurrent connections**: Any connection to a workspace (e.g. SSH, web terminal, `coder_app`)
- **provisioner daemons**: Number of processes responsible for workspace builds, per coderd replica. 
    
    ```text
    2 coderd replicas * 30 provisioner daemons = 30 max concurrent workspace builds
    ```
    
- **scaletest**: Our scale-testing utility, built into the `coder` command line.

## General recommendations

### Concurrent workspace builds

Workspace builds are CPU-intensive, as it relies on Terraform and the various [Terraform providers](https://registry.terraform.io/browse/providers). When tested with our [kubernetes](https://github.com/coder/coder/tree/main/examples/templates/kubernetes) template, `coderd` will consume roughly 8 cores per 30 concurrent workspace builds. For effective provisioning, our helm chart prefers to schedule [one coderd replica per-node](https://github.com/coder/coder/blob/main/helm/values.yaml#L110-L121).

To support 120 concurrent workspace builds, for example:

> - Create a cluster/nodepool with three 8-core nodes (AWS: `t3.2xlarge` GCP: `e2-highcpu-8`)

Reviewer: To match the 120 concurrent builds calculation
```suggestion
- Create a cluster/nodepool with four 8-core nodes (AWS: `t3.2xlarge` GCP: `e2-highcpu-8`)
```

> We regularly scale-test Coder with our [scale testing utility](#scaletest-utility). The same utility can be used in your own environment for insights on how Coder performs with your specific templates, images, etc.

Reviewer 2: ```suggestion
We scale-test Coder with the [same utility](#scaletest-utility) that can be used in your environment for insights into how Coder scales with your infrastructure.
```

## General concepts

Coder runs workspace operations in a queue. The number of concurrent builds will be limited to the number of provisioner daemons across all coderd replicas.

```text
2 coderd replicas * 30 provisioner daemons = 60 max concurrent workspace builds

> ```

Reviewer 2: I might put this below the definitions.

- **concurrent connections**: Any connection to a workspace (e.g. SSH, web terminal, `coder_app`)
- **provisioner daemons**: Coder runs one workspace build per provisioner daemon. One coderd replica can host many daemons

> - **scaletest**: Our scale-testing utility, built into the `coder` command line.

Reviewer 3: nit: People who are not familiar with Coder internals may find it hard to imagine it. You can consider adding a picture later. I know that they can skim through linked documents, but it saves a few minutes of life :)

Author: would love to do this down the road.

```text
2 coderd replicas * 30 provisioner daemons = 60 max concurrent workspace builds
```

## Infrastructure recommendations

### Concurrent workspace builds

Workspace builds are CPU-intensive, as it relies on Terraform. Various [Terraform providers](https://registry.terraform.io/browse/providers) have different resource requirements. When tested with our [kubernetes](https://github.com/coder/coder/tree/main/examples/templates/kubernetes) template, `coderd` will consume roughly 8 cores per 30 concurrent workspace builds. For effective provisioning, our helm chart prefers to schedule [one coderd replica per-node](https://github.com/coder/coder/blob/main/helm/values.yaml#L110-L121).

To support 120 concurrent workspace builds, for example:

> - Create a cluster/nodepool with four 8-core nodes (AWS: `t3.2xlarge` GCP: `e2-highcpu-8`)

Reviewer 3: nit: with a total of 4 nodes, 8-core each

- Run coderd with 4 replicas, 30 provisioner daemons each. (`CODER_PROVISIONER_DAEMONS=30`)
- Ensure Coder's [PostgreSQL server](../../admin/configure.md#postgresql-database) can use up to 1.5 cores

## Recent scale tests

> | Environment        | Users | Concurrent builds | Concurrent connections (SSH) | Concurrent connections (web) | Last tested  |

Reviewer 3: Do you think we can mention here the used version of Coder? Probably, you can deduce it from date (`Last tested`).

> | ------------------ | ----- | ----------------- | ---------------------------- | ---------------------------- | ------------ |

Reviewer 3: It might be relevant for users to know what's the total time of the scale test. If scale tests performed on their infra take twice as much time, it will be a hint for them to start digging.

Author: We can make sure to include the time next time we run these tests.

Reviewer 3: You may want to add a column already, so the next person will not forget about measuring it.

| Kubernetes (GKE)   | 1200  | 120               | 10,000                       | 10,000                       | Jan 10, 2022 |
| Docker (Single VM) | 500   | 50                | 10,000                       | 10,000                       | Dec 20, 2022 |

## Scale testing utility

Since Coder's performance is highly dependent on the templates and workflows you support, we recommend using our scale testing utility against your own environments.

The following command will run our scale test against your own Coder deployment. You can also specify a template name and any parameter values.

```sh
coder scaletest create-workspaces \
    --count 1000 \
    --template "kubernetes" \

>     --concurrency 0 \

Reviewer 3: You may want to explain here what `concurrency 0` and `cleanup-concurrency 0` mean.

Author: @deansheather what do you think about making `0` the default value for these? I was a bit confused why this defaults to 1.

`concurrency 0` means the client will request 1000 workspaces at once (no limit). `concurrency 200` means that the client will request a max of 200 requests at a given time.

Reviewer 4: Sure. We could make it a required flag so admins consciously opt-in to infinite rather than infinite by default.

    --cleanup-concurrency 0 \
    --parameter "home_disk_size=10" \
    --run-command "sleep 2 && echo hello"

# Run `coder scaletest create-workspaces --help` for all usage
```

> To avoid potential outages and orphaned resources, we recommend running scale tests on a secondary "staging" environment.

The test does the following:

- create `n` workspaces
- establish SSH connection to each workspace
- run `sleep 3 && echo hello` on each workspace via the web terminal
- close connections, attempt to delete all workspaces

> - return results (e.g. `99 succeeded, 1 failed to connect`)

Reviewer 3: nit: numeric ordering? It's easier to refer to points in the future.
